[ML] Add backend for getting ML autoscaling resources

hendrikmuhs opened the pull request:
add transport action to retrieve autoscaling metrics for ML autoscaling

Note: `removeNodeMemoryInBytes` not implemented yet, this will be part of a follow up

hendrikmuhs committed:
add an internal action to get required autoscaling resources

hendrikmuhs committed:
remove unused logger

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

hendrikmuhs committed:
rename internal name

hendrikmuhs committed:
rename classes to fit internal name

hendrikmuhs committed:
rename AutoscalingResources to MlAutoscalingStats

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java:
```diff
@@ -0,0 +1,160 @@
+            } else {
+                modelMemoryBytesSum += mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId);
+                nodesWithRunningJobs.add(task.getExecutorNode());
+                minNodes = 1;
```
Should we set `minNodes` to 1 even if the only anomaly detection jobs are unassigned?

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java:
```diff
@@ -0,0 +1,160 @@
+                    .getTaskParams()
+                    .getThreadsPerAllocation();
+
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, unassignedProcessors);
```
We don't need every allocation to be on a single node - in theory each allocation could be put on a different node. So the extra processors required on a single node should only be the threads per allocation.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java:
```diff
@@ -0,0 +1,160 @@
+                    .getThreadsPerAllocation();
+
+                // min(3, max(number of allocations over all deployed models), as this is always 3
+                minNodes = Math.min(3, Math.max(minNodes, assignedProcessors));
```
This formula should use number of allocations rather than assigned processors. For example, 1 allocation with 3 threads per allocation should only require 1 node and not 3.

Also, should `minNodes` be set in the same way even if the only models are awaiting allocation?

droberts195 reviewed (commented):
The overall structure of this looks good. I just found a few nits in the details.

One more complication is that we assume the first 30MB of ML native memory is lost to loading shared library code. This is why lots of the calculations in the current autoscaling use variables that include `ExcludingPerNodeOverhead` in their names, for example: https://github.com/elastic/elasticsearch/blob/fb4addab65c5badc1623cc04e8895d0fb00d807b/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NativeMemoryCapacity.java#L208

Implicitly _all_ your variables in this PR are "excluding per node overhead".

But the way this is dealt with in current code is overly complex. We can achieve the same effect by just deducting 30MB from the amounts of ML native memory the autoscaling controller thinks each node has, which is a trivial tweak to the autoscaling controller code in place of lots of complexity in the Elasticsearch side code.

All we need is a comment saying that the calculations are all in terms of native memory _excluding_ per node overhead, and that's dealt with on the autoscaling controller side.

hendrikmuhs committed:
improve structure and code docs

hendrikmuhs committed:
- add dfa accounting
- don't count CPUs for dfa and ad
- simplify minNodes calculation
- base single node CPUs for trained models only on threads

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java:
```diff
@@ -0,0 +1,160 @@
+                    mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId)
+                );
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, 1);
+                ++extraProcessors;
```
Currently DFA doesn't require any CPU either. The current logic for processors is in: https://github.com/elastic/elasticsearch/blob/d704f21ef9a54ee504221879a47c98b628383d5a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlProcessorAutoscalingDecider.java#L114-L137

Currently that only considers trained models.

(I don't think it's necessary to carry over the 0.1 processor requirement for low priority models. The memory requirement should ensure low priority models get a node to run on.)

droberts195 replied:
It's probably a good idea to add comments in both the new and old autoscaling code to say that if we ever change this decision so that jobs require processors then both places should be changed.

hendrikmuhs committed:
report perNodeMemoryOverheadInBytes as part of the response

hendrikmuhs commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportGetMlAutoscalingStats.java:
```diff
@@ -0,0 +1,95 @@
+                ActionListener.wrap(autoscalingResources -> listener.onResponse(new Response(autoscalingResources)), listener::onFailure)
+            );
+        } else {
+            mlMemoryTracker.refresh(
```
I enforced the timeout and return a 408 in case. The outer API will make proper use of that and not return the ML metrics(but the other ones) and the autoscaler should interpret the ML-part as a no-op.

That way we still return something if it is quick or if the refresh was about to finish from a previous run.

hendrikmuhs commented:
Regarding the per node overhead: I added a new field `perNodeMemoryOverheadInBytes`, so that we report that back to the autoscaler and the autoscaler can handle it. I think that's the right contract: the plugin owns the value, but the autoscaler has to deal with it. Arguably this value will probably never change, however this way it is technical clear and the price is a few extra bytes to be serialized and deserialized.

hendrikmuhs committed:
adjust comment

droberts195 reviewed (approved):
LGTM

hendrikmuhs commented:
Thanks @droberts195,

I am sure this is not the final version, but given this we can iterate and start end2end testing.
